### PR TITLE
feat: Quick fixes for typos in instruction and macro names added to the code actions

### PR DESCRIPTION
**Pull request by slavek-kucera**

### CODE REVIEW

File: utils/test/bk_tree_test.cpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+struct abs_value
+{
+    auto operator()(T a, T b) const noexcept { return (size_t)std::abs(a - b); }
+};
+
+struct lv_dist_sv
+{
+    auto operator()(std::string_view l, std::string_view r) const noexcept { return levenshtein_distance(l, r); }
+};
+
```
Comment (jirimosinger) [inline utils/test/bk_tree_test.cpp:33]: Maybe add a test for looking up a value while the tree is empty?

File: language_server/src/lsp/feature_language_features.cpp [hunk 2]
```diff
@@ -402,13 +452,50 @@ json feature_language_features::document_symbol_list_json(
...
-    auto document_uri = params["textDocument"]["uri"].get<std::string>();
+    auto document_uri = extract_document_uri(params);
 
     const auto limit = 5000LL;
     auto symbol_list = ws_mngr_.document_symbol(document_uri.c_str(), limit);
 
     response_->respond(id, "", document_symbol_list_json(symbol_list));
 }
 
+void feature_language_features::opcode_suggestion(const json& id, const json& params)
```
Comment (jirimosinger) [inline language_server/src/lsp/feature_language_features.cpp:463]: Do you plan to add the suggestions to the completion list as well? I think it might be even more handy there since the user could see the more options.

E.g. When I type `AAA` I could be glad to see other options than `ADATA` like the suggested ones `LAA`, `LA` and others

File: parser_library/src/workspace_manager_impl.h [hunk 3]
```diff
@@ -262,6 +262,21 @@ class workspace_manager::impl final : public diagnosable_impl
...
+        const std::string& document_uri, const std::string& opcode, bool extended)
+    {
+        auto suggestions =
+            ws_path_match(document_uri)
+                .make_opcode_suggestion(utils::resource::resource_location(document_uri), opcode, extended);
+
+        std::vector<opcode_suggestion> res;
+
+        for (auto&& [suggestion, distance] : suggestions)
+            res.emplace_back(opcode_suggestion { make_continuous_sequence(std::move(suggestion)), distance });
```
Comment (jirimosinger) [inline parser_library/src/workspace_manager_impl.h:275]: This comes from SonarCloud:

```suggestion
            res.emplace_back(make_continuous_sequence(std::move(suggestion)), distance);
```

File: utils/include/utils/bk_tree.h [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            size_t* id;
+            ~restore_t()
+            {
+                if (id)
+                    *id = invalid;
+            }
+        } restore { node_id };
+
+        *node_id = m_nodes.size();
+        const auto& new_node = m_nodes.emplace_back(node_t { distance, invalid, invalid, std::forward<U>(value) });
```
Comment (jirimosinger) [inline utils/include/utils/bk_tree.h:120]: This comes from SonarCloud
```suggestion
        const auto& new_node = m_nodes.emplace_back(distance, invalid, invalid, std::forward<U>(value));
```

File: parser_library/src/workspace_manager_impl.h (hunk 3 above)
Comment (slavek-kucera) [reply parser_library/src/workspace_manager_impl.h:275]: no go - clang 12-14 does not support aggregate_type(...) construction.

File: utils/include/utils/bk_tree.h (hunk 4 above)
Comment (slavek-kucera) [reply utils/include/utils/bk_tree.h:120]: no go - clang 12-14 issue

File: language_server/src/lsp/feature_language_features.cpp (hunk 2 above)
Comment (slavek-kucera) [reply language_server/src/lsp/feature_language_features.cpp:463]: that sound like a good idea, maybe a follow up PR?

File: clients/vscode-hlasmplugin/src/hlasmCodeActionsProvider.ts [hunk 5]
```diff
@@ -14,9 +14,69 @@
...
+function unique(a: string[]) {
+    return [...new Set(a)];
+}
+
+async function gatherOpcodeSuggestions(opcodes: string[], client: vscodelc.BaseLanguageClient, uri: vscode.Uri): Promise<OpcodeSuggestionList> {
+    await client.onReady();
+    const suggestionsResponse = await client.sendRequest<OpcodeSuggestionResponse>("textDocument/$/opcode_suggestion", {
+        textDocument: { uri: uri.toString() },
+        opcodes: opcodes,
+        extended: false,
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply clients/vscode-hlasmplugin/src/hlasmCodeActionsProvider.ts:41]: originally there was an idea to have a "show more suggestions"-like code action, I don't really want to remove it at this moment and the internal implementation could be e.g. used e.g. for the fuzzy completion list idea...

File: parser_library/src/workspaces/processor_group.cpp [hunk 6]
```diff
@@ -54,6 +56,57 @@ processor_group::processor_group(const std::string& pg_name,
...
+    }
+    else
+        m_suggestions.emplace();
+
+    for (const auto& l : m_libs)
+    {
+        for (auto&& f : l->list_files())
+        {
+            if (f.size() > suggestion_limit)
+                continue;
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply parser_library/src/workspaces/processor_group.cpp:75]: I don't follow, this is just to skip files with e.g. 200 character in their name - I'd say the absolute limit is 63 char due to hlasm grammar rules, practical limit is 8 really... so 32 should be a big enough.

File: parser_library/src/workspaces/processor_group.h [hunk 7]
```diff
@@ -48,11 +50,19 @@ class processor_group : public diagnosable_impl
...
+
+    std::vector<std::pair<std::string, size_t>> suggest(std::string_view s, bool extended);
+
 private:
     std::vector<std::unique_ptr<library>> m_libs;
     std::string m_pg_name;
     config::assembler_options m_asm_opts;
     std::vector<preprocessor_options> m_prep_opts;
+
+    static constexpr size_t suggestion_limit = 32;
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply parser_library/src/workspaces/processor_group.h:63]: static variable, so definitely not `m_`... I usually expect capitalized names like that to be a preprocessor symbol. So I prefer the name as it is now.

File: parser_library/src/workspaces/workspace.cpp [hunk 8]
```diff
@@ -336,6 +339,100 @@ const processor_group& workspace::get_proc_grp_by_program(const utils::resource:
...
+    const utils::resource::resource_location& file, std::string_view opcode, bool extended)
+{
+    std::vector<std::pair<std::string, size_t>> result;
+
+    asm_option opts;
+    if (const auto* pgm = m_configuration.get_program(file); pgm)
+    {
+        auto& proc_grp = m_configuration.get_proc_grp_by_program(*pgm);
+        proc_grp.update_asm_options(opts);
+        pgm->asm_opts.apply(opts);
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply parser_library/src/workspaces/workspace.cpp:420]: override_options, apply_override_options?

File: utils/include/utils/bk_tree.h (hunk 4 above)
Comment (jirimosinger) [reply utils/include/utils/bk_tree.h:120]: Right...

File: language_server/src/lsp/feature_language_features.cpp (hunk 2 above)
Comment (jirimosinger) [reply language_server/src/lsp/feature_language_features.cpp:463]: Sure!

File: clients/vscode-hlasmplugin/src/hlasmCodeActionsProvider.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jirimosinger) [reply clients/vscode-hlasmplugin/src/hlasmCodeActionsProvider.ts:41]: Ok, I just wanted to know the reasoning for this. Thx!

File: parser_library/src/workspaces/processor_group.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/workspaces/processor_group.h:63]: Alright

File: parser_library/src/workspaces/processor_group.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/workspaces/processor_group.cpp:75]: Right... I got confused and thought that `f` represents a library instead of a file name...

So, maybe we can change `f` to `filename`?

**Review by jirimosinger — APPROVED** [review]
